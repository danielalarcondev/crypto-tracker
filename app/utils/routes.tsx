export enum Routes {
	"ASSETS" = "/assets",
	"DEFAULT" = "/tracker",
	"TRACKER" = "/tracker"
}

export const getDefaultRoute = () => Routes.DEFAULT;