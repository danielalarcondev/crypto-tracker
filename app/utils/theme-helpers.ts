
interface ThemeHelper {
	getIsDarkTheme: () => boolean;
    isDarkThemeConfigured: () => boolean;
    // eslint-disable-next-line no-unused-vars
    setIsDarkTheme: (shouldSetToDark: boolean) => void;
}

const IS_DARK_THEME_KEY = 'IS_DARK_THEME';

export const isDarkThemeConfigured = () => localStorage?.getItem(IS_DARK_THEME_KEY) !== null;
export const getIsDarkTheme = () => Boolean(localStorage?.getItem(IS_DARK_THEME_KEY));

export const setIsDarkTheme = (shouldSetToDark: boolean) => {
    localStorage?.setItem(IS_DARK_THEME_KEY, shouldSetToDark.toString());
};

const themeLocalStorage: ThemeHelper = {
    getIsDarkTheme,
    isDarkThemeConfigured,
    setIsDarkTheme
};

export default themeLocalStorage;
 
