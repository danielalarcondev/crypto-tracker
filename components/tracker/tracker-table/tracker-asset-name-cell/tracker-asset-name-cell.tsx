'use client';

import React, { SyntheticEvent } from 'react';
import Image from 'next/image';

import { TableCell } from 'flowbite-react';

interface TrackerAssetNameCellProps {
	name: string;
	symbol: string;
}

export const defaultNameCelltUrl = '/logo.svg';

export default function TrackerAssetNameCell({ name, symbol }: TrackerAssetNameCellProps) {

    const iconUrl = `https://assets.coincap.io/assets/icons/${symbol.toLocaleLowerCase()}@2x.png`;

    const onImageError = (event: SyntheticEvent<HTMLImageElement, Event>) => {
        (event.target as HTMLImageElement).src = defaultNameCelltUrl;
        (event.target as HTMLImageElement).removeAttribute('srcset');
    };

    return (
        <TableCell 
            className={[
                'flex', 
                'w-fit', 
                'max-w-32', 
                'gap-2', 
                'whitespace-nowrap', 
                'font-medium', 
                ['text-gray-900', 'dark:text-white'].join(' ')]
                .join(' ')
            }
        >
            <Image
                data-testid="tracker-table-name-image"
                width={32}
                height={32}
                src={iconUrl}
                alt="icon"
                onError={onImageError}
                className={[
                    'w-8', 
                    'h-8'
                	].join(' ')
                }
            />
            <div 
                className={[
                    'flex', 
                    'flex-col', 
                    'self-center'
                	].join(' ')
                }
            >
                <span 
                    className={[
                        'font-bold'
                    ].join(' ')
                    }
                >
                    {name}
                </span>
                <small 
                    className={[
                        'text-neutral-400', 
                        'pl-0.5'
                    ].join(' ')
                    }
                >
                    {symbol}
                </small>
            </div>
        </TableCell>
    );
}
