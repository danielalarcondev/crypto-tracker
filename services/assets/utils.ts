import { GetAssetsResponse } from '@common-types/assets';

export const DEFAULT_ASSETS_LIMIT = 100;
export const DEFAULT_ASSETS_OFFSET = 0;

export const DEFAULT_ERROR_CODE = 500;
export const DEFAULT_ERROR_MESSAGE = 'Something went wrong';
export const URL_ERROR_CODE = 404;
export const URL_ERROR_MESSAGE = 'Server could not be reached. Try again later';

export const defaultErrorResponse: GetAssetsResponse = {
    data: [],
    timestamp: Date.now(),
    error: { 
        errorMessage: DEFAULT_ERROR_MESSAGE, 
        httpError: DEFAULT_ERROR_CODE
    }
};

export const urlErrorResponse: GetAssetsResponse = {
    data: [],
    timestamp: Date.now(),
    error: { 
        errorMessage: URL_ERROR_MESSAGE,
        httpError: URL_ERROR_CODE
    }
};
