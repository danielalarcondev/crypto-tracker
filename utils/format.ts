export const formatLargeNumber = (number: string) => {

    const formatter: Intl.NumberFormat = new Intl.NumberFormat('en-US', {
        notation: 'compact',
        compactDisplay: 'short',
        minimumFractionDigits: 2,
        maximumFractionDigits: 2
    });

    return formatter.format(Number(number));
};

export const formatPercentageNumber = (percentage: string) => {

    const TWO_DECIMALS = 2;
    const number = Number(percentage);

    if (isNaN(number)) {
        return `${percentage}%`;
    }

    return `${number.toFixed(TWO_DECIMALS)}%`;
};
