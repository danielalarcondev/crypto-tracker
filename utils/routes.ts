export enum Routes {
	"ASSETS" = "/assets",
	"DEFAULT" = "/tracker",
	"TRACKER" = "/tracker"
}
