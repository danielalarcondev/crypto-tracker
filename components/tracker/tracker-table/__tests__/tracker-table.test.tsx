import * as React from 'react';
import '@testing-library/jest-dom/jest-globals';

import { fireEvent, render, screen } from '@testing-library/react';
import { expect, it, describe } from '@jest/globals';
import TrackerTable from '@components/tracker/tracker-table/tracker-table';
import { defaultNameCelltUrl } from '@components/tracker/tracker-table/tracker-asset-name-cell/tracker-asset-name-cell';
import { mockedAssets } from '@tests/utils';

const FIRST_INDEX = 0;
const firstAsset = mockedAssets[FIRST_INDEX];

describe('TrackerTable: ', () => {

    it('should render correctly', () => {

        const { container } = render(<TrackerTable assets={mockedAssets} />);
        expect(container).toMatchSnapshot();
    });

    describe('Name cell: ', () => {
        it('should render correct image', () => {

            render(<TrackerTable assets={[firstAsset]} />);
	
            const img = screen.getByTestId('tracker-table-name-image');
            expect(img.getAttribute('src')).toBe(`http://localhost/_next/image?url=https%3A%2F%2Fassets.coincap.io%2Fassets%2Ficons%2F${firstAsset.symbol.toLocaleLowerCase()}%402x.png&w=64&q=75`);
        });
	
        it('should render fallback image on error', () => {
	
            render(<TrackerTable assets={[firstAsset]} />);
	
            const img = screen.getByTestId('tracker-table-name-image');
			
            fireEvent.error(img);
            expect(img.getAttribute('src')).toBe(defaultNameCelltUrl);
        });
    });
});
