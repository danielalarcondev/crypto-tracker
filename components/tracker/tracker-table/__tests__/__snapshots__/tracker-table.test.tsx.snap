// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackerTable:  should render correctly 1`] = `
<div>
  <div
    class="overflow-x-auto rounded-lg"
    data-testid="tracker-table"
  >
    <div
      class="relative"
      data-testid="table-element"
    >
      <div
        class="absolute left-0 top-0 -z-10 h-full w-full drop-shadow-md dark:bg-neutral-950"
      />
      <table
        class="w-full text-left text-sm text-neutral-500 dark:text-neutral-400"
      >
        <thead
          class="group/head text-xs uppercase text-neutral-700 dark:text-neutral-100"
        >
          <tr>
            <th
              class="px-4 md:px-6 py-4 bg-neutral-100 dark:bg-neutral-600 hidden xl:table-cell w-24"
            >
              #
            </th>
            <th
              class="px-4 md:px-6 py-4 bg-neutral-100 dark:bg-neutral-600 w-fit"
            >
              Name
            </th>
            <th
              class="px-4 md:px-6 py-4 bg-neutral-100 dark:bg-neutral-600 w-32"
            >
              Price
            </th>
            <th
              class="px-4 md:px-6 py-4 bg-neutral-100 dark:bg-neutral-600 hidden xs:table-cell w-32"
            >
              24h %
            </th>
            <th
              class="px-4 md:px-6 py-4 bg-neutral-100 dark:bg-neutral-600 hidden xl:table-cell w-32"
            >
              Market cap
            </th>
            <th
              class="px-4 md:px-6 py-4 bg-neutral-100 dark:bg-neutral-600 hidden lg:table-cell w-32"
            >
              Supply
            </th>
            <th
              class="px-4 md:px-6 py-4 bg-neutral-100 dark:bg-neutral-600 hidden sm:table-cell w-32"
            >
              Volume 24h
            </th>
          </tr>
        </thead>
        <tbody
          class="group/body dark:text-neutral-200 divide-y"
        >
          <tr
            class="group/row bg-white dark:bg-neutral-800"
            data-testid="table-row-element"
          >
            <td
              class="md:px-6 neutral-50 dark:bg-neutral-800 p-4 w-24 hidden xl:table-cell"
            >
              1
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 flex w-fit max-w-32 gap-2 whitespace-nowrap font-medium text-gray-900 dark:text-white"
            >
              <img
                alt="icon"
                class="w-8 h-8"
                data-nimg="1"
                data-testid="tracker-table-name-image"
                decoding="async"
                height="32"
                loading="lazy"
                src="http://localhost/_next/image?url=%2Floading-icon.gif&w=64&q=75"
                srcset="/_next/image?url=%2Floading-icon.gif&w=32&q=75 1x, /_next/image?url=%2Floading-icon.gif&w=64&q=75 2x"
                style="color: transparent;"
                width="32"
              />
              <div
                class="flex flex-col self-center"
              >
                <span
                  class="font-bold"
                >
                  Bitcoin
                </span>
                <small
                  class="text-neutral-400 pl-0.5"
                >
                  BTC
                </small>
              </div>
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 w-32"
            >
              $6.93K
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 hidden xs:table-cell w-32"
            >
              -0.81%
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 hidden xl:table-cell w-32"
            >
              119.15B
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 hidden lg:table-cell w-32"
            >
              17.19M
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 hidden sm:table-cell w-32"
            >
              2.93B
            </td>
          </tr>
          <tr
            class="group/row bg-white dark:bg-neutral-800"
            data-testid="table-row-element"
          >
            <td
              class="md:px-6 neutral-50 dark:bg-neutral-800 p-4 w-24 hidden xl:table-cell"
            >
              2
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 flex w-fit max-w-32 gap-2 whitespace-nowrap font-medium text-gray-900 dark:text-white"
            >
              <img
                alt="icon"
                class="w-8 h-8"
                data-nimg="1"
                data-testid="tracker-table-name-image"
                decoding="async"
                height="32"
                loading="lazy"
                src="http://localhost/_next/image?url=%2Floading-icon.gif&w=64&q=75"
                srcset="/_next/image?url=%2Floading-icon.gif&w=32&q=75 1x, /_next/image?url=%2Floading-icon.gif&w=64&q=75 2x"
                style="color: transparent;"
                width="32"
              />
              <div
                class="flex flex-col self-center"
              >
                <span
                  class="font-bold"
                >
                  Ethereum
                </span>
                <small
                  class="text-neutral-400 pl-0.5"
                >
                  ETH
                </small>
              </div>
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 w-32"
            >
              $404.97
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 hidden xs:table-cell w-32"
            >
              -0.09%
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 hidden xl:table-cell w-32"
            >
              40.97B
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 hidden lg:table-cell w-32"
            >
              101.16M
            </td>
            <td
              class="px-4 md:px-6 py-4 neutral-50 dark:bg-neutral-800 hidden sm:table-cell w-32"
            >
              1.03B
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
